File: leetCode/string/easy/quiz125/java/SolutionV2.java
package string.easy.quiz125.java;

public class SolutionV2 {
    public static void main(String[] args) {
        boolean answer = new SolutionV2().isPalindrome("A man, a plan, a canal: Panama");
        System.out.println(answer);
    }

    public boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            while (left < right && !isAlphanumeric(toLowerCase(s.charAt(left)))) {
                left++;
            }
            while (left < right && !isAlphanumeric(toLowerCase(s.charAt(right)))) {
                right--;
            }

            if (toLowerCase(s.charAt(left)) != toLowerCase(s.charAt(right))) {
                return false;
            }

            left++;
            right--;
        }

        return true;
    }

    private boolean isAlphanumeric(char ch) {
        return ch >= 'a' && ch <= 'z' ||
                ch >= '0' && ch <= '9';
    }

    private char toLowerCase(char ch) {
        if (ch >= 'A' && ch <= 'Z') {
            return (char) (ch + 32);
        } else {
            return ch;
        }
    }
}

File: leetCode/string/easy/quiz125/kotlin/SolutionV2.kt
package string.easy.quiz125.kotlin

fun main() {
    SolutionV2().isPalindrome("A man, a plan, a canal: Panama").apply {
        println(this)
    }
}

class SolutionV2 {
    fun isPalindrome(s: String): Boolean {
        var left = 0
        var right = s.lastIndex

        while (left < right) {
            while (left < right && !isAlphanumericOrDigit(s[left].toLowerCase())) {
                left++
            }

            while (left < right && !isAlphanumericOrDigit(s[right].toLowerCase())) {
                right++
            }

            if (s[left].toLowerCase() != s[right].toLowerCase()) {
                return false
            }

            left++
            right--
        }

        return true
    }

    private fun isAlphanumericOrDigit(c: Char): Boolean {
        return (c in 'a'..'z') || (c in '0'..'9')
    }

    private fun Char.toLowerCase(): Char {
        return if (this in 'A'..'Z') this + 32 else this
    }
}
